src/cliffordsynthesis/encoding/TwoQubitEncoder.cpp

```diff
@@ -0,0 +1,251 @@
+                       tvars->singleQubitRChange(pos, qubit, gate),
+                       LogicTerm(0, static_cast<std::int16_t>(S)));
+    splitXorR(change, pos);
+  }
+}
+
+void TwoQubitEncoder::assertGatesImplyTransform(
+    const std::size_t pos, const std::size_t qubit,
+    const std::vector<TransformationFamily>& transformations) {
+  const auto& singleQubitGates = vars.gS[pos / 2];
+  for (const auto& [transformation, gates] : transformations) {
+    auto gateOr = LogicTerm(false);
+    for (const auto& gate : gates) {
+      gateOr = gateOr || singleQubitGates[gateToIndex(gate)][qubit];
+    }
+    lb->assertFormula(LogicTerm::implies(gateOr, transformation));
+  }
+}
+
+void encoding::TwoQubitEncoder::assertPauliGateConstraints(
+    const std::size_t pos) {
+  for (std::size_t q = 0U; q < N; ++q) {
+    for (const auto gate : PAULI_QUBIT_GATES) {
+      const auto& change =
+          LogicTerm::ite(vars.gP[gateToIndex(gate)][q],
+                         tvars->singleQubitRChange(pos, q, gate),
+                         LogicTerm(0, static_cast<std::int16_t>(S)));
+
+      splitXorR(change, pos);
+    }
+    lb->assertFormula(tvars->x[pos][q] == tvars->x[pos + 1][q]);
+    lb->assertFormula(tvars->z[pos][q] == tvars->z[pos + 1][q]);
+  }
+}
+
+void TwoQubitEncoder::splitXorR(const logicbase::LogicTerm& changes,
+                                std::size_t                 pos) {
+  auto&             xorHelper = xorHelpers[pos];
+  const std::string hName =
+      "h_" + std::to_string(pos) + "_" + std::to_string(xorHelper.size());
+  DEBUG() << "Creating helper variable for RChange XOR " << hName;
+  const auto n = static_cast<std::int16_t>(S);
+  xorHelper.emplace_back(lb->makeVariable(hName, CType::BITVECTOR, n));
+  if (xorHelper.size() == 1) {
+    lb->assertFormula(xorHelper.back() == changes);
+  } else {
+    lb->assertFormula(xorHelper.back() ==
+                      (xorHelper[xorHelpers[pos].size() - 2] ^ changes));
+  }
+}
+
+void encoding::TwoQubitEncoder::assertTwoQubitGateConstraints(
+    const std::size_t pos) {
+  const auto& twoQubitGates = vars.gC[pos / 2];
+  for (std::size_t ctrl = 0U; ctrl < N; ++ctrl) {
+    for (std::size_t trgt = 0U; trgt < N; ++trgt) {
+      if (ctrl == trgt) {
+        const auto changes = createIdentityConstraintOnTQG(pos, ctrl);
+        lb->assertFormula(
+            LogicTerm::implies(twoQubitGates[ctrl][ctrl], changes));
+        splitXorR(tvars->singleQubitRChange(pos, ctrl, qc::OpType::None), pos);
+      } else {
+        const auto changes = createTwoQubitGateConstraint(pos, ctrl, trgt);
+        lb->assertFormula(
+            LogicTerm::implies(twoQubitGates[ctrl][trgt], changes));
+        DEBUG() << "Asserting CNOT on " << ctrl << " and " << trgt;
+      }
+    }
+  }
+}
+
+LogicTerm
+encoding::TwoQubitEncoder::createIdentityConstraintOnTQG(std::size_t pos,
+                                                         std::size_t ctrl) {
```

Comment:
This does not have to be a method. It is simple enough to be inlined and is only called once.